### PR TITLE
MGMT-13045: Enhancement document for node labeling

### PR DESCRIPTION
**Pull request by ori-amizur**
## List all the issues related to this PR

- [ ] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [x] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md
/cc @filanov 
/cc @avishayt

### CODE REVIEW

File: docs/enhancements/support-node-labeling.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+### Goals
+- For ZTP flow: add capability to add and update labels on BMH which will be transferred 
+to agent CR.
+- For boot it yourself flow: add capability to add and update labels to agent CR after agent
+  has been created.
+
+### Non goals
+- Single way to set node labels before agent creation both for ZTP flow, and boot it yourself flow.
+- REST API support.  Node labeling will not be supported under REST API.
```
Comment (filanov) [inline docs/enhancements/support-node-labeling.md:30]: It's already supported, but only in day1 https://github.com/openshift/assisted-service/blob/2b0f66c4cdcdd2dc8556d7f65b7f3e0eac0c958a/swagger.yaml#L4116-L4119

File: docs/enhancements/support-node-labeling.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+## Proposal
+
+There are 2 implementation options: Using labels, or adding dedicated spec field.
+
+### Using labels
```
Comment (filanov) [inline docs/enhancements/support-node-labeling.md:36]: because we are talking about a specific case of adding labels to nodes i don't think that we should use the labels on the agent for that.

File: docs/enhancements/support-node-labeling.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+Labels can be applied on every CR.  So there is no known limitation to add a label on agent.
+
+For the ZTP flow, there should a pattern for transferring an annotation from the BMH to a label  
+in the corresponding agent.  The pattern is actually the prefix of the annotation.
+It should be **agent-install.openshift.io.user.role**.  These annotations will be transferred to 
+to labels on the corresponding agent CR.
+
+### Adding dedicated spec field in agent CR
+Additional field **userRoles** will be added to the spec
```
Comment (filanov) [inline docs/enhancements/support-node-labeling.md:45]: suggesting to be more generic and call it node labels
Roles is just one of the use cases here.

File: docs/enhancements/support-node-labeling.md (hunk 1 above)
Comment (ori-amizur) [reply docs/enhancements/support-node-labeling.md:30]: OK, so it is non goal.

File: docs/enhancements/support-node-labeling.md (hunk 3 above)
Comment (ori-amizur) [reply docs/enhancements/support-node-labeling.md:45]: OK

File: docs/enhancements/support-node-labeling.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+There was a requirement to support node roles during installation.  These roles 
+are extra roles, which are used by the user and not by any Openshift components.
+These labels should be applied on agent CR which represents a node in a spoke cluster.
+
+## Motivation
+
+Currently, nodes can be assigned with a single role which can be one of two values [ master, worker].
+The may want to add additional roles.  These roles are for the user usage only.
+
+### Goals
```
Comment (romfreiman) [inline docs/enhancements/support-node-labeling.md:22]: what about day2?

File: docs/enhancements/support-node-labeling.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+- "@oamizur"
+  creation-date: 2022-12-20
+  last-updated: 2022-12-20
+---
+
+# Support node labeling during installation
+
+## Summary
+
+There was a requirement to support node roles during installation.  These roles
```
Comment (filanov) [inline docs/enhancements/support-node-labeling.md:13]: roles is just a use case here, i suggest to have summary talk about the general use case, which is defining labels on the agent that will result in the nodes on the spoke cluster after the installation

File: docs/enhancements/support-node-labeling.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+section of the agent CR.  This field will be a string list.
+
+For the ZTP flow, the annotations having the format **agent-install.openshift.io.user.role/[key]=[value]**,
+the value part will be transferred to the **userRoles** field as additional entry. 
+
+### User Stories
+
+#### Story 1
+
+Let the user define roles specific to the business domain in addition to
```
Comment (romfreiman) [inline docs/enhancements/support-node-labeling.md:55]: what will it transfer to in OCP? what will it affect?

**Comment by romfreiman** [comment]
@ori-amizur im looking into ROSA demo (2:45 min) - do see there node labels. Lets understand how it's being implemented with hive?
https://www.youtube.com/watch?v=gAMr3sI5bdY&t=169s

**Comment by avishayt** [comment]
I agree with @filanov .  We have node_labels in the REST API, let's add it to the ACI Spec and a corresponding annotation on the BMH.

**Comment by romfreiman** [comment]
I dont understand who/how this will help? It wont reduce any reboot, neither wont control mcp that the host will be part of.

**Comment by romfreiman** [comment]
There is a work that was dont by @tsorya in OCP 4.12 that can help, but it does not really need a new API i beleive.

https://github.com/openshift/machine-config-operator/blob/master/templates/worker/01-worker-kubelet/on-prem/units/kubelet.service.yaml#L29

File: docs/enhancements/support-node-labeling.md (hunk 5 above)
Comment (tsorya) [reply docs/enhancements/support-node-labeling.md:13]: You can't set custom role before node joins

File: docs/enhancements/support-node-labeling.md (hunk 2 above)
Comment (tsorya) [reply docs/enhancements/support-node-labeling.md:36]: There is already field for node_labels, please just use it

File: docs/enhancements/support-node-labeling.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+For the ZTP flow, there should a pattern for transferring an annotation from the BMH to a label  
+in the corresponding agent.  The pattern is actually the prefix of the annotation.
+It should be **agent-install.openshift.io.user.role**.  These annotations will be transferred to 
+to labels on the corresponding agent CR.
+
+### Adding dedicated spec field in agent CR
+Additional field **userRoles** will be added to the spec 
+section of the agent CR.  This field will be a string list.
+
+For the ZTP flow, the annotations having the format **agent-install.openshift.io.user.role/[key]=[value]**,
```
Comment (tsorya) [inline docs/enhancements/support-node-labeling.md:48]: Not specific for role

**Comment by tsorya** [comment]
> There is a work that was dont by @tsorya in OCP 4.12 that can help, but it does not really need a new API i beleive.
> 
> https://github.com/openshift/machine-config-operator/blob/master/templates/worker/01-worker-kubelet/on-prem/units/kubelet.service.yaml#L29

It allows to set labels when kubelet start though it doesn't allow to set roles and some more labels that are blocked by kubelet code

**Comment by romfreiman** [comment]
> > There is a work that was dont by @tsorya in OCP 4.12 that can help, but it does not really need a new API i beleive.
> > https://github.com/openshift/machine-config-operator/blob/master/templates/worker/01-worker-kubelet/on-prem/units/kubelet.service.yaml#L29
> 
> It allows to set labels when kubelet start though it doesn't allow to set roles and some more labels that are blocked by kubelet code

right, but this api wont help neither

**Comment by oourfali** [comment]
@ori-amizur regarding ROSA, iirc it is done via syncsets OCM defined in hive. Perhaps that can be leveraged via assisted in ACM as well.

File: docs/enhancements/support-node-labeling.md (hunk 4 above)
Comment (oourfali) [reply docs/enhancements/support-node-labeling.md:22]: iiuc the RFE says they are doing it today at day-2 which is something they prefer to avoid.
Let's sit with Jose and Ian to understand the details.

File: docs/enhancements/support-node-labeling.md (hunk 2 above)
Comment (filanov) [reply docs/enhancements/support-node-labeling.md:36]: but if it will exist in the DB we will need to take into account that the controller will use that

File: docs/enhancements/support-node-labeling.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+the value part will be transferred to the **userRoles** field as additional entry. 
+
+### User Stories
```
Comment (filanov) [inline docs/enhancements/support-node-labeling.md:51]: @imiller0 maybe you can provide additional user stories? 
it will make the requirement clearer. 
For now i'm not sure that fully understand the requirement or if we are even able to execute it with the existing tools.

File: docs/enhancements/support-node-labeling.md (hunk 4 above)
Comment (ori-amizur) [reply docs/enhancements/support-node-labeling.md:22]: It will be done through the Kube API, so the controller has access to the spoke nodes.

File: docs/enhancements/support-node-labeling.md (hunk 5 above)
Comment (ori-amizur) [reply docs/enhancements/support-node-labeling.md:13]: These roles do not have operational meaning from our POV.  These will be user defined attributes for their own use.

File: docs/enhancements/support-node-labeling.md (hunk 7 above)
Comment (ori-amizur) [reply docs/enhancements/support-node-labeling.md:48]: OK.  We will discuss with @imiller0 to clarify the requirements.

File: docs/enhancements/support-node-labeling.md (hunk 5 above)
Comment (tsorya) [reply docs/enhancements/support-node-labeling.md:13]: This is depends on what you mean by roles. I talked about k8s role and you can set them only after node joins k8s cluster

File: docs/enhancements/support-node-labeling.md (hunk 4 above)
Comment (tsorya) [reply docs/enhancements/support-node-labeling.md:22]: I assume you mean service will have and not controller right?

File: docs/enhancements/support-node-labeling.md (hunk 2 above)
Comment (tsorya) [reply docs/enhancements/support-node-labeling.md:36]: i assume you need to set them in cr and then in db as all other things we do, right? I just mean that field in db already exists and we need to set it in kubeapi flow

File: docs/enhancements/support-node-labeling.md (hunk 7 above)
Comment (tsorya) [reply docs/enhancements/support-node-labeling.md:48]: It doesn't really mean what are specific requirements as we are going to set labels, what user means but this label doesn't really matter, it is still called label

File: docs/enhancements/support-node-labeling.md (hunk 6 above)
Comment (ori-amizur) [reply docs/enhancements/support-node-labeling.md:85]: It will be transferred to labels on nodes

File: docs/enhancements/support-node-labeling.md (hunk 2 above)
Comment (ori-amizur) [reply docs/enhancements/support-node-labeling.md]: Yes

File: docs/enhancements/support-node-labeling.md (hunk 6 above)
Comment (ori-amizur) [reply docs/enhancements/support-node-labeling.md:85]: to labels on nodes.
It may effect the user role which may cause attaching to different MCP

File: docs/enhancements/support-node-labeling.md [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+## Proposal
+
+### ZTP flow
+Labels can be applied on every CR.  So there is no known limitation to add a label on agent.
+
+For the ZTP flow, there should a pattern for transferring an annotation from the BMH to a label  
+in the corresponding agent.  The pattern is actually the prefix of the annotation.
+It should be **node-label**.  The suffix of the annotation will be transferred 
+to labels on the corresponding agent CR.
```
Comment (avishayt) [inline docs/enhancements/support-node-labeling.md:54]: This section contradicts the next section ("Node labels on an agent CR").  Here you say that we are "transferring an annotation from the BMH to a label in the corresponding agent" but below you say that annotations "will be transferred to the dedicated field nodeLabels".

I think the next section is correct and you can erase this one.  We should not automatically copy labels between resources but rather have it explicit in the API.

File: docs/enhancements/support-node-labeling.md [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+#### Day 2 nodes
+The flow can be supported only when working with Kubernetes API, since direct access to
+the cluster is required.  In this case, setting node labels should be done by a 
+Kuberenetes controller (probably agent controller).
+
+In addition, for day 2 nodes we need to support setting custom manifests.  This functionality is
+already supported for day 1 clusters by the assisted controller.  In order to allow adding new
+MCP to existing day 2 clusters, this functionality needs to be supported by a Kubernetes controller - 
+probably the cluster-deployment controller.
```
Comment (avishayt) [inline docs/enhancements/support-node-labeling.md:84]: Adding MCPs to existing clusters feels out of scope for Assisted.

File: docs/enhancements/support-node-labeling.md [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+
+For the ZTP flow, the annotations having the prefix **node-label** in BMH will be transferred
+to the dedicated field **nodeLabels** without the **node-label** prefix.
+
+If boot it yourself flow is in use, nodeLabels should be set directly on the agent CR.
+
+### Moving labels to nodes
+
+#### Day 1 cluster
+When installing day 1 cluster, the implementation should make use of the **nodeLabels** field
```
Comment (avishayt) [inline docs/enhancements/support-node-labeling.md:71]: Why nodeLabels and not machineConfigPoolName?

Comment (tsorya) [reply docs/enhancements/support-node-labeling.md:71]: I assume both. User will need to set mcp + role in nodelables

Comment (avishayt) [reply docs/enhancements/support-node-labeling.md:71]: @ori-amizur , this enhancement needs more details.
- The title indicates that the end goal is labeling nodes, but that's different from what the use case is.
- It would help to have a description of the steps required to do this with UPI and then what the steps would be with CIM after this proposal.  Here is a partial list of questions in my head after reading this: Does the user need to create an MCP?  How?  Does the user also need to use the machineConfigPoolName API?  How?  What is that API for and what is this label for?  If we need both machineConfigPoolName and label, can the existing API just add the label?

Comment (ori-amizur) [reply docs/enhancements/support-node-labeling.md:71]: Node labels can be used for many purposes - not necessarily roles.  This enhancement is for node labels, which custom roles are specific use-case for node labels.

Comment (ori-amizur) [reply docs/enhancements/support-node-labeling.md:71]: @avishayt although in terms of user story, MCP and MC are related to node labels, creation of MC and MCP are done by creating manifests.  This is already supported by AI for day1 installation.
As I indicated in the document, we should also support it for day2

File: docs/enhancements/support-node-labeling.md (hunk 10 above)
Comment (ori-amizur) [reply docs/enhancements/support-node-labeling.md:84]: OK.

File: docs/enhancements/support-node-labeling.md (hunk 9 above)
Comment (ori-amizur) [reply docs/enhancements/support-node-labeling.md:54]: OK

File: docs/enhancements/support-node-labeling.md (hunk 11 above)
Comment (ori-amizur) [reply docs/enhancements/support-node-labeling.md:71]: MCP name is another field that already exists. No need to add it.

**Comment by avishayt** [comment]
The document is missing an end-to-end flow.  I'm a user that wants to use a custom role - what APIs do I need to use?

**Comment by ori-amizur** [comment]
> The document is missing an end-to-end flow. I'm a user that wants to use a custom role - what APIs do I need to use?

1.  Add node label
2. Add manifests: MachineConfigPool and possibly MachineConfig
